R/sampler-adaptive.R

```diff
@@ -0,0 +1,308 @@
+    u <- rng$random_real(1)
+    density_next <- model$density(pars_next)
+    
+    accept_prob <- min(1, exp(density_next - state$density))
+
+    accept <- u < accept_prob
+    if (accept) {
+      state$pars <- pars_next
+      state$density <- density_next
+    }
+
+    internal$iteration <- internal$iteration + 1
+    internal$history_pars <- rbind(internal$history_pars, state$pars) 
+    if (internal$iteration > adapt_end) {
+      internal$scaling_history <- c(internal$scaling_history, internal$scaling)
+      return(state)
+    }
+    
+    if (internal$iteration > pre_diminish) {
+      internal$scaling_weight <- internal$scaling_weight + 1
+    }
+    
+    is_replacement <- 
+      check_replacement(internal$iteration, forget_rate, forget_end)
+    if (is_replacement) {
+      pars_remove <- internal$history_pars[internal$included[1L], ]
+      internal$included <- c(internal$included[-1L], internal$iteration)
+    } else {
+      pars_remove <- NULL
+      internal$included <- c(internal$included, internal$iteration)
+      internal$weight <- internal$weight + 1
+    }
+    
+    internal$scaling <- 
+      update_scaling(internal$scaling, internal$scaling_weight, accept_prob,
+                     internal$scaling_increment, min_scaling, acceptance_target, 
+                     log_scaling_update)
+    internal$scaling_history <- c(internal$scaling_history, internal$scaling)
+    internal$autocorrelation <- update_autocorrelation(
+      state$pars, internal$weight, internal$autocorrelation, pars_remove)
+    internal$mean <- update_mean(state$pars, internal$weight, internal$mean,
+                                 pars_remove)
+    internal$vcv <- update_vcv(internal$mean, internal$autocorrelation,
+                               internal$weight)
+
+    state
+  }
+
+  finalise <- function(state, model, rng) {
+    out <- as.list(internal)
+    out[c("autocorrelation", "mean", "vcv", "weight", "included",
+          "scaling_history", "scaling_weight", "scaling_increment")]
+  }
+
+  mcstate_sampler("Adaptive Metropolis-Hastings",
+                  initialise,
+                  step,
+                  finalise)
+}
+
+
+calc_scaling_increment <- function(n_pars, acceptance_target,
+                                   log_scaling_update) {
+  if (log_scaling_update) {
+    A <- -qnorm(acceptance_target / 2)
+    
+    scaling_increment <- 
+      (1 - 1 / n_pars) * (sqrt(2 * pi) * exp(A^2 / 2)) / (2 * A) + 
+      1 / (n_pars * acceptance_target * (1 - acceptance_target))
+  } else {
+    scaling_increment <- 1 / 100
+  }
+  
+  scaling_increment
+}
+
+
+qp <- function(x) {
+  outer(x, x)
+}
+
+
+calc_proposal_vcv <- function(scaling, vcv, weight, initial_vcv,
+                              initial_vcv_weight) {
+  n_pars <- nrow(vcv)
+  
+  weighted_vcv <-
+    ((weight - 1) * vcv + (initial_vcv_weight + n_pars + 1) * initial_vcv) /
+    (weight + initial_vcv_weight + n_pars + 1)
+  
+  2.38^2 / n_pars * scaling^2 * weighted_vcv
```

Comment:
why is `scaling` squared here? in the algorithm they do not square it

Reply:
There is a typo in the algorithm in the paper, it should be squared! It is stated correctly in the text:

![image](https://github.com/mrc-ide/mcstate2/assets/47318334/39aa377c-50c1-4f86-9c98-eedf45161991)